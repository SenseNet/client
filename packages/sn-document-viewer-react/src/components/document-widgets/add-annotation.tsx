import { createStyles, makeStyles, Theme } from '@material-ui/core'
import { ChatSharp } from '@material-ui/icons'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddAnnnotationClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Annotation
 */
export const AddAnnotationWidget: React.FC<{ classes?: AddAnnnotationClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  const pageRotation =
    viewerState.rotation?.find((rotation) => rotation.pageNum === viewerState.activePage)?.degree || 0

  return (
    <ToggleBase
      disabled={!permissions.canEdit || pageRotation !== 0}
      classes={classes}
      active={viewerState.activeShapePlacing === 'annotation'}
      title={localization.addAnnotation}
      setValue={(value) => viewerState.updateState({ activeShapePlacing: value ? 'annotation' : 'none' })}>
      <ChatSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.activeShapePlacing === 'annotation' })}
      />
    </ToggleBase>
  )
}
