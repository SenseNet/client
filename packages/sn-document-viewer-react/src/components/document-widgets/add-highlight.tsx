import { createStyles, makeStyles, Theme } from '@material-ui/core'
import ChatBubbleOutlineSharp from '@material-ui/icons/ChatBubbleOutlineSharp'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddRHightlightClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Hihgligh
 */
export const AddHighlightWidget: React.FC<{ classes?: AddRHightlightClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  return (
    <ToggleBase
      disabled={!permissions.canEdit}
      classes={classes}
      isVisible={viewerState.isPlacingHighlight}
      title={localization.addHighlight}
      setValue={(value) => viewerState.updateState({ isPlacingHighlight: value })}>
      <ChatBubbleOutlineSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.isPlacingHighlight })}
      />
    </ToggleBase>
  )
}
