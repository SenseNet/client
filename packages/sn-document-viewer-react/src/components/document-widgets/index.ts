export * from './add-annotation'
export * from './add-highlight'
export * from './add-redaction'
export * from './document-title-pager'
export * from './download-widget'
export * from './print-widget'
export * from './rotate-active-pages'
export * from './rotate-document'
export * from './rotate-widget'
export * from './save-widget'
export * from './share-widget'
export * from './toggle-base'
export * from './toggle-comments'
export * from './toggle-redaction'
export * from './toggle-shapes'
export * from './toggle-thumbnails'
export * from './toggle-watermark'
export * from './zoom-in-out'
