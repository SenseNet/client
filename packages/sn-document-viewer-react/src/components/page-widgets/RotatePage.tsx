import { PreviewImageData } from '@sensenet/client-core'
import IconButton from '@material-ui/core/IconButton'
import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'
import React, { useCallback, useState } from 'react'
import { useLocalization, usePreviewImage } from '../../hooks'
import { Dimensions } from '../../services'

/**
 * The amount of rotation in degrees
 */
export const ROTATION_AMOUNT = 90

/**
 * Visibility of the rotation related buttons
 */
export enum ROTATION_MODE {
  clockwise,
  anticlockwise,
  all,
}

/**
 * Defined the component's own properties
 */
export interface RotatePageProps {
  page: PreviewImageData
  viewPort: Dimensions
  zoomRatio: number
}

export const RotatePageWidget: React.FC<RotatePageProps> = (props) => {
  const localization = useLocalization()

  const [pageIndex] = useState(props.page.Index)
  const image = usePreviewImage(pageIndex)

  const rotatePageLeft = useCallback(() => {
    image.rotate(-ROTATION_AMOUNT)
  }, [image])

  const rotatePageRight = useCallback(() => {
    image.rotate(ROTATION_AMOUNT)
  }, [image])

  return (
    <div
      style={{
        position: 'absolute',
        zIndex: 1,
        top: 0,
        right: 0,
        filter: 'drop-shadow(0 0 3px white) drop-shadow(0 0 5px white) drop-shadow(0 0 9px white)',
      }}>
      <IconButton onClick={rotatePageLeft} title={localization.rotatePageLeft}>
        <RotateLeft />
      </IconButton>

      <IconButton onClick={rotatePageRight} title={localization.rotatePageRight}>
        <RotateRight />
      </IconButton>
    </div>
  )
}
