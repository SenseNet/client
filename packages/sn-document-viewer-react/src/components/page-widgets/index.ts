export * from './RotatePage'
export * from './Shape'
export * from './Shapes'
export * from './style'
