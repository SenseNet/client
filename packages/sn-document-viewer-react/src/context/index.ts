export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
