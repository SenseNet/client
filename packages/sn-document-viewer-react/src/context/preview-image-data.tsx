import React, { useCallback, useEffect, useState } from 'react'
import { PreviewImageData } from '@sensenet/client-core'
import {
  useDocumentData,
  useDocumentPermissions,
  useDocumentViewerApi,
  useViewerSettings,
  useViewerState,
} from '../hooks'
import { applyShapeRotations, ImageUtil } from '../services'

export const PreviewImageDataContext = React.createContext<{
  imageData: PreviewImageData[]
  rotateImages: (indexes: number[], amount: number) => void
}>({ imageData: [], rotateImages: () => undefined })

export const PreviewImageDataContextProvider: React.FC = (props) => {
  const viewerSettings = useViewerSettings()
  const api = useDocumentViewerApi()
  const { documentData, updateDocumentData } = useDocumentData()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()
  const [previewImages, setPreviewImages] = useState<PreviewImageData[]>([])

  useEffect(() => {
    const abortController = new AbortController()
    ;(async () => {
      try {
        setPreviewImages([])
        const images = await api.getExistingPreviewImages({
          document: documentData,
          version: viewerSettings.version || '',
          showWatermark: viewerState.showWatermark,
          abortController,
        })
        setPreviewImages(images)
      } catch (error) {
        if (!abortController.signal.aborted) {
          throw error
        }
      }
    })()

    return () => abortController.abort()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    api,
    documentData.idOrPath,
    documentData.hostName,
    viewerSettings.documentIdOrPath,
    viewerSettings.version,
    viewerState.showWatermark,
  ])

  const rotateImages = useCallback(
    (imageIndexes: number[], amount: number) => {
      if (!permissions.canEdit) {
        console.warn(`No permission to edit!`)
        return
      }
      const newImages = previewImages.map((img) => {
        const newImg = { ...img }
        if (imageIndexes.indexOf(newImg.Index) >= 0) {
          const newAngle =
            ImageUtil.normalizeDegrees(((newImg.Attributes && newImg.Attributes.degree) || 0) + (amount % 360)) % 360
          newImg.Attributes = {
            ...newImg.Attributes,
            degree: newAngle,
          }
        }

        updateDocumentData({
          shapes: {
            annotations: applyShapeRotations(documentData.shapes.annotations, amount, img),
            highlights: applyShapeRotations(documentData.shapes.highlights, amount, img),
            redactions: applyShapeRotations(documentData.shapes.redactions, amount, img),
          },
        })

        return newImg
      })
      setPreviewImages(newImages)
      viewerState.updateState({ hasChanges: true })
    },
    [documentData, updateDocumentData, permissions.canEdit, previewImages, viewerState],
  )

  return (
    <PreviewImageDataContext.Provider
      value={{
        imageData: previewImages,
        rotateImages,
      }}>
      {props.children}
    </PreviewImageDataContext.Provider>
  )
}
