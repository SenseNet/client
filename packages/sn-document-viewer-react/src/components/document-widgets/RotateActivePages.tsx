import React, { useCallback } from 'react'
import IconButton from '@material-ui/core/IconButton'
import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'

import { useLocalization, usePreviewImages, useViewerState } from '../../hooks'
import { ROTATION_AMOUNT } from '../page-widgets/RotatePage'
/**
 * Component that allows active page rotation
 */
export const RotateActivePagesWidget: React.FC = () => {
  const localization = useLocalization()
  const viewerState = useViewerState()
  const images = usePreviewImages()

  const rotateDocumentLeft = useCallback(() => {
    images.rotateImages(viewerState.activePages, -ROTATION_AMOUNT)
  }, [images, viewerState.activePages])

  const rotateDocumentRight = useCallback(() => {
    images.rotateImages(viewerState.activePages, ROTATION_AMOUNT)
  }, [images, viewerState.activePages])

  /**
   * renders the component
   */
  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton
        color="inherit"
        title={localization.rotatePageLeft}
        onClick={() => rotateDocumentLeft()}
        id="RotateActiveLeft">
        <RotateLeft />
      </IconButton>
      <IconButton
        color="inherit"
        title={localization.rotatePageRight}
        onClick={() => rotateDocumentRight()}
        id="RotateActiveRight">
        <RotateRight />
      </IconButton>
    </div>
  )
}
