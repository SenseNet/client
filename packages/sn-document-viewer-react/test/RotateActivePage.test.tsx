import IconButton from '@material-ui/core/IconButton'
import { mount } from 'enzyme'
import React from 'react'
import { RotateActivePagesWidget } from '../src/components/document-widgets/RotateActivePages'
import { ROTATION_AMOUNT, ROTATION_MODE } from '../src/components/page-widgets/RotatePage'
import { PreviewImageDataContext } from '../src/context/preview-image-data'
import { examplePreviewImageData } from './__Mocks__/viewercontext'

describe('RotateActivePage component', () => {
  it('Should render without crashing', () => {
    const wrapper = mount(<RotateActivePagesWidget />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should render without crashing with mode param', () => {
    const wrapper = mount(<RotateActivePagesWidget mode={ROTATION_MODE.clockwise} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should render without crashing with mode param', () => {
    const wrapper = mount(<RotateActivePagesWidget mode={ROTATION_MODE.anticlockwise} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should render without crashing with mode param', () => {
    const wrapper = mount(<RotateActivePagesWidget mode={ROTATION_MODE.all} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('RotateLeft should trigger a rotate to left', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ rotateImages } as any}>
        <RotateActivePagesWidget />
      </PreviewImageDataContext.Provider>,
    )

    wrapper.find(IconButton).first().simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], -ROTATION_AMOUNT)
    expect(rotateImages).toBeCalled()
  })

  it('RotateRight should trigger a rotate to right', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ rotateImages } as any}>
        <RotateActivePagesWidget />
      </PreviewImageDataContext.Provider>,
    )

    wrapper.find(IconButton).last().simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], ROTATION_AMOUNT)
    expect(rotateImages).toBeCalled()
  })

  it('RotateLeft should trigger a rotate to left', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ rotateImages } as any}>
        <RotateActivePagesWidget mode={ROTATION_MODE.anticlockwise} />
      </PreviewImageDataContext.Provider>,
    )

    wrapper.find(IconButton).first().simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], -ROTATION_AMOUNT)
    expect(rotateImages).toBeCalled()
  })

  it('RotateRight should trigger a rotate to right', () => {
    const rotateImages = jest.fn()
    const wrapper = mount(
      <PreviewImageDataContext.Provider value={{ rotateImages } as any}>
        <RotateActivePagesWidget mode={ROTATION_MODE.clockwise} />
      </PreviewImageDataContext.Provider>,
    )

    wrapper.find(IconButton).last().simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], ROTATION_AMOUNT)
    expect(rotateImages).toBeCalled()
  })
})
