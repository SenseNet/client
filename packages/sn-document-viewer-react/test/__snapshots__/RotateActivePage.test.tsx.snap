// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotateActivePage component Should render without crashing 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <WithStyles(ForwardRef(IconButton))
    color="inherit"
    id="RotateActiveLeft"
    onClick={[Function]}
    title="rotateDocumentLeft"
  >
    <RotateLeftIcon />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(IconButton))
    color="inherit"
    id="RotateActiveRight"
    onClick={[Function]}
    title="rotateDocumentRight"
  >
    <RotateRightIcon />
  </WithStyles(ForwardRef(IconButton))>
</div>
`;
